core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.java

```diff
@@ -638,10 +639,26 @@ public EnvelopePart slice(
 
     /** Returns a new EnvelopePart, where all positions are shifted by positionDelta */
     public EnvelopePart copyAndShift(double positionDelta) {
-        var newPositions = new double[positions.length];
-        for (int i = 0; i < positions.length; i++)
-            newPositions[i] = positions[i] + positionDelta;
-        return new EnvelopePart(new HashMap<>(attrs), newPositions, speeds, timeDeltas);
+        var newPositions = new ArrayList<Double>();
+        var newSpeeds = new ArrayList<Double>();
+        var newTimeDeltas = new ArrayList<Double>();
```

Comment:
You can use [hppc's DoubleArrayList](http://carrotsearch.github.io/hppc/releases/0.7.1/api/com/carrotsearch/hppc/DoubleArrayList.html) for better performance and decreased memory usage